### DISCUSSION TITLE
CDC from planetscale

### DISCUSSION THREAD

**Discussion by amcclosky** [planetscale/discussion#119] [Feature Requests] [answered]
Is there already a solution for or plans to implement a feature that would allow change data capture with planetscale? i.e. deliver an event to my application with the changed data when a row is updated.

**Comment by johnyerhot** [comment]
:wave: @amcclosky - so something like a webhook whose payload contains the schema diff that fires when a database branch is merged?

We have been discussing adding webhooks that fire on various events quite a bit internally. I can't make any promises on if or when, though.

**Reply by amcclosky** [reply]
That would be cool but I was thinking something more along the lines of a web-hook that fires when I insert or update a row in a table.

Alternatively, having mysql binlog access would also work. I haven't attempted to verify myself yet but does planetscale allow a client to connect to a database for replication?

**Reply by jgreet** [reply]
We don't currently expose MySQL binlogs from PlanetScale databases.

**Reply by ethangclark** [reply]
binlogs or a hook that fires with created/updated/deleted row information would really set planetscale apart for reactive projects.

Current workaround is polling on `updatedAt` columns.

**Reply by amcclosky** [reply]
> We don't currently expose MySQL binlogs from PlanetScale databases.

Thanks for the response @jgreet. Completely understandable limitation at this stage of the product. I'd assume exposing binlogs or equivalent is on the roadmap at some point in the future though.

**Reply by chinmay-apps** [reply]
> binlogs or a hook that fires with created/updated/deleted row information would really set planetscale apart for reactive projects.
> 
> Current workaround is polling on `updatedAt` columns.

Hello team, we could not sync delete operation, whereas create and update operations are getting properly synced up. Any pointers towards getting this will help my team a lot. It will really set planetscale a way ahead, as @ethangclark has rightly point out.

I am just a newbie at planetscale and any help in this regard is greatly appreciated.
I subscribed to the basic plan at planetscale, but not sure where to post my query regarding Airbyte connector for dealing with syncing of delete operations.

**Comment by LouisIV** [comment]
It would be amazing if planet scale could expose the VStream from the VTGate. This is the difference between me hosting Vitess myself and using planet scale. 😢 

A first party example showing how to integrate it with something like debeizum would be 🚀 

Let me know if there is anything I can do to move this along

**Reply by mscoutermarsh** [reply]
Debezium is now available: https://planetscale.com/docs/integrations/debezium

**Comment by mscoutermarsh** [comment, marked as answer]
PlanetScale Connect is now available: https://planetscale.com/blog/extract-load-and-transform-your-data-with-planetscale-connect

We also have a Debezium connector: https://planetscale.com/docs/integrations/debezium

**Reply by streamkap** [reply]
Any update on exposing binlogs?

**Reply by tiagoefreitas** [reply]
This is not an answer, Connect is ETL with Airbyte that is not CDC at all. CDC means realtime updates on every transaction.
Airbyte launches Docker Containers to sync data between source and destination and can run minimum every 5 minutes.

In CDC the destination is not event a database just a message.

**Reply by mscoutermarsh** [reply]
Update: we now have a debezium connector available for PlanetScale https://planetscale.com/docs/integrations/debezium

**Reply by pkit** [reply]
Same shit again, it's not a CDC. lol

**Reply by mscoutermarsh** [reply]
@pkit Can you share what's missing for you with the debezium connector?

**Reply by paulfdudley** [reply]
If folks are interested in a managed provider of the debezium connector, we just launched the [Streamkap PlanetScale CDC connector](https://docs.streamkap.com/docs/planetscale#/) on our platform. We now support streaming from PlanetScale to a bunch of destinations like Snowflake, Databricks, BigQuery, ClickHouse, Motherduck, and more.

**Comment by CanRau** [comment]
Insert, update, delete webhook(s) would be incredible 🤓
This would be very helpful to build queue (like) data structures directly integrated as part of the main database 😻

**Comment by Tang8330** [comment]
Any updates on exposing binlogs?

**Reply by jgreet** [reply]
Nothing to announce regarding binlogs. We'll post an update here if that changes.

**Comment by chinmay-apps** [comment]
Anybody, who have used this Airbyte connector and has been able to sync delete operations as well, please help me.

**Reply by chinmay-apps** [reply]
Planetscale provides "Full refresh Overwrite" as a refresh option to sync up delete operations. But this can not be accepted as a solution as this will have very bad performance when number of records in source database is huge. Any other suggestions team?

**Reply by JohnGemstone** [reply]
I think for now we have to use a pub/sub service like Ably or host our own service using something like [Soketi](https://docs.soketi.app/). These dont directly connect to pscale but you use them along side, mirroring events and seperating them into their own logical channels clients can subscribe to using webhooks.

**Comment by etdej** [comment]
Hi there,
Any news on supporting CDC ? Is it prioritised in your roadmap ?
Thanks,

**Reply by mscoutermarsh** [reply]
We just released a debezium connector for PlanetScale: https://planetscale.com/docs/integrations/debezium

**Comment by andrao** [comment]
This is not official support, but for those working in TS/JS I've created a package, [planetscale-stream-ts](https://github.com/andrao/planetscale-stream-ts), which provides access to the Vitess VStream, a CDC log.

Under the hood it uses the [psdbconnect gRPC API](https://github.com/planetscale/psdb/tree/main/proto-src/psdbconnect), which is used for e.g. the [Connect Airbyte adapter](https://github.com/planetscale/airbyte-source/tree/main). This API **is in alpha**, so use with caution! This response from PlanetScale Support is encouraging, however:

> I received confirmation that the v1alpha1 API will be around for quite a while since it has been adopted for use in a number of places so deprecating it would likely only occur in a very slow / controlled manner at this point.

**Reply by etdej** [reply]
Thanks ! Does this handle deletes properly ? The Connect Airbyte adapter does not (which limits its usability very much). I am struggling to understand where this limitation comes from.

**Reply by andrao** [reply]
> Thanks ! Does this handle deletes properly ? The Connect Airbyte adapter does not (which limits its usability very much). I am struggling to understand where this limitation comes from. 

@etdej it does indeed! See the screencap in the README there.

**Reply by etdej** [reply]
Yeah 🥳

**Comment by nkolly** [comment]
For those looking for easy access to the PlanetScale CDC log, I am a YC swe and I created [planetscale-cdc.com](https://planetscale-cdc.com) to deliver real-time webhook events for every insert/update/delete operation. It utilizes the Vstream api and we handle all the complexity that comes with that. More integrations coming soon. The current alternatives cause data loss or are more setup and cost than is worth it at our stage. Using [planetscale-cdc.com](https://planetscale-cdc.com) has unlocked real-time leaderboards and exact audit logs at my startup. Hope it can help you too!
